### PR TITLE
BUG: stop pickle from making copies of builtin dtypes.

### PR DESCRIPTION
**Pull request by J-Sand**
Fixes #4317. Also add a minor clarification to the C API docs.

### CODE REVIEW

File: doc/source/reference/c-api.types-and-structures.rst [hunk 1]
```diff
@@ -355,13 +355,14 @@ PyArrayDescr_Type
 
 .. c:member:: PyObject *PyArray_Descr.fields
 
-    If this is non-NULL, then this data-type-descriptor has fields
+    For built-in data types, this is set to ``Py_None``. Otherwise,
```
Comment (charris) [inline doc/source/reference/c-api.types-and-structures.rst]: Can it be NULL? If so, what does that mean.

Comment (J-Sand) [inline doc/source/reference/c-api.types-and-structures.rst]: Yes, for non-builtin dtypes without any fields. I guess the wording needs improving?

File: numpy/core/src/multiarray/descriptor.c [hunk 2]
```diff
@@ -2360,9 +2360,16 @@ arraydescr_reduce(PyArray_Descr *self, PyObject *NPY_UNUSED(args))
     PyObject *ret, *mod, *obj;
     PyObject *state;
     char endian;
-    int elsize, alignment;
+    int elsize, alignment, builtin;
 
-    ret = PyTuple_New(3);
+    builtin = self->fields == Py_None;
+    if (builtin) {
```
Comment (charris) [inline numpy/core/src/multiarray/descriptor.c]: Function documentation needs fixing for this case.

File: numpy/core/src/multiarray/descriptor.c [hunk 3]
```diff
@@ -2391,7 +2398,10 @@ arraydescr_reduce(PyArray_Descr *self, PyObject *NPY_UNUSED(args))
         }
         obj = PyUString_FromFormat("%c%d",self->kind, elsize);
     }
-    PyTuple_SET_ITEM(ret, 1, Py_BuildValue("(Nii)", obj, 0, 1));
+    PyTuple_SET_ITEM(ret, 1, Py_BuildValue("(Nii)", obj, 0, !builtin));
```
Comment (charris) [inline numpy/core/src/multiarray/descriptor.c]: Is it clear that byte order does not apply? What if the pickle is unpikled on a machine with different endianess?

File: doc/source/reference/c-api.types-and-structures.rst (hunk 1 above)
Comment (charris) [inline doc/source/reference/c-api.types-and-structures.rst]: Agree about the wording. I was trying to figure out what all the permutations were ;)

File: numpy/core/src/multiarray/descriptor.c (hunk 3 above)
Comment (charris) [inline numpy/core/src/multiarray/descriptor.c]: Hmm, builtins do have endianess. Does builtin have cross platform implications? I suppose that depends on how things get unpickled. Or does the endianess have anything to do with the builtin designation? Perhaps I am confusing builtin with native.

Comment (J-Sand) [inline numpy/core/src/multiarray/descriptor.c]: The third element of ret? It's set right at the end of the function, it contains the extra state information needed for arraydescr_setstate to rebuild the dtype (for non-builtin dtypes). The reduce function is allowed to just return a 2-tuple if the third element isn't needed.

Comment (charris) [inline numpy/core/src/multiarray/descriptor.c]:
```
In [10]: d = ones(3, '>f8').dtype

In [11]: d.isnative
Out[11]: False

In [12]: d.isbuiltin
Out[12]: 0

In [13]: d = ones(3, '<f8').dtype

In [14]: d.isnative
Out[14]: True

In [15]: d.isbuiltin
Out[15]: 1
```
So it looks like `builtin` may not be portable.

**Comment by charris** [comment]
Made me wonder how builtin is detected. Is there a dictionary or list someplace?

**Comment by charris** [comment]
There is a ` _builtin_descrs` that looks like an array indexed by type.

**Comment by J-Sand** [comment]
Yeah, `PyArray_DescrFromType` retrieves them from `_builtin_descrs`.

I realised I also needed to be a bit more careful about extended and datetime types, which also have some extra state set. 'U' has elsize/alignment, which I guess should at least vary depending on narrow/wide build, and the datetime types have their metadata. Though the built-in datetime dtypes, 'm' and 'M', seem kind of broken anyway. Trying to make an array with dtype 'M' results in:

    TypeError: Invalid datetime unit "generic" in metadata

You also get the same exception when you try and unpickle either `np.dype('m')` or `np.dtype('M')`. The dtypes with time units attached aren't built-in anyway.

Huh, every Travis build is segfaulting...

**Comment by J-Sand** [comment]
Now I'm getting lots of unrelated test failures on 3.6-dev, it looks like [this CPython commit](https://hg.python.org/cpython/rev/aaee06743c61) has broken numpy's deprecation warnings?

**Comment by seberg** [comment]
Probably breaking my suppress_warning context, but don't have time to look at it right now.

**Comment by charris** [comment]
I filed a bug with python. There was a rc1 release yesterday and it seems that tests of the form
```
    with warnings.catch_warnings(record=True) as w:
        warnings.simplefilter('always')
        ...
```
fail because nothing gets recorded in `w`.

**Comment by charris** [comment]
Apropos builtins, I got to wondering how  the long type `"l"` gets handled because is varies in size between windows and linux. I wish Python had never heard of the long type ;)

**Comment by J-Sand** [comment]
It looks like 'l' gets converted into something more specific when you create a dtype anyway:

    >>> np.dtype('l')
    dtype('int64')
    >>> np.dtype('>l')
    dtype('>i8')

[The docs](https://docs.scipy.org/doc/numpy/reference/arrays.scalars.html#arrays-scalars-built-in) seem to say that all of the built-in types are built-in everywhere except for the long double/long complex types (but pickles containing those are already non-cross-platform, and I don't think my changes should make them any worse). I checked that `int64` and `int32` are both built-in in 64-bit Linux and Windows anyway.

**Comment by charris** [comment]
Might be worth checking specifically with the `"l"` type. The 'int64' and '>i8' are many to one descriptors.

**Comment by J-Sand** [comment]
Sorry, I think you've gone over my head. Do you mean that even things like '>i8' differ from one platform to another?

**Comment by charris** [comment]
`'>i8'` should have the same precision and endianess on all platforms, but might have different underlying C type, for instance on one platform it may be `long`, on another `long long`. So it depends on which info the unpickle uses. You probably have a better idea than me at this point if it is safe or not, I may fretting about a possibility that doesn't exist.

File: numpy/core/src/multiarray/descriptor.c [hunk 4]
```diff
@@ -2357,27 +2359,41 @@ arraydescr_reduce(PyArray_Descr *self, PyObject *NPY_UNUSED(args))
      * arraydescr_setstate.
     */
     const int version = 4;
-    PyObject *ret, *mod, *obj;
+    PyObject *ret, *obj;
     PyObject *state;
     char endian;
-    int elsize, alignment;
+    int elsize, alignment, no_state;
+    static PyObject *np_dtype;
```
Comment (charris) [inline numpy/core/src/multiarray/descriptor.c]: Need to initialize to `NULL`.

**Comment by charris** [comment]
Could you squash the commits into one? I don't know if we have any testing hardware for big endian these days, but it would be nice to check that this works on it. @matthew-brett Do you access to such hardware?

**Comment by charris** [comment]
@J-Sand Be good to finish this up.

**Comment by OmerJog** [comment]
Is there an update on this PR?

**Comment by charris** [comment]
@J-Sand Going to close this. If you want to pursue this, please open a new PR.

**Pull request closed without merging**